jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -1657,286 +1528,345 @@ ecma_builtin_string_prototype_object_split (ecma_value_t this_arg, /**< this arg
+        separator = ecma_copy_value (separator_to_string_val);
 
-        JERRY_ASSERT (ecma_is_value_true (put_comp));
+        ECMA_FINALIZE (separator_to_string_val);
+      }
 
-        should_return = true;
+      const ecma_string_t *this_to_string_p = ecma_get_string_from_value (this_to_string_val);
 
-        ecma_deref_ecma_string (zero_str_p);
-      }
-      else /* if (!ecma_is_value_undefined (arg1)) */
+      /* 11. */
+      if (ecma_string_is_empty (this_to_string_p) && ecma_is_value_empty (ret_value))
       {
-        /* 8. */
-        ecma_value_t separator = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
+        bool should_return = false;
 
-        if (ecma_is_value_object (arg1)
-            && ecma_object_class_is (ecma_get_object_from_value (arg1), LIT_MAGIC_STRING_REGEXP_UL))
+        if (separator_is_regexp)
         {
-          separator = ecma_copy_value (arg1);
+#ifndef CONFIG_DISABLE_REGEXP_BUILTIN
+          ecma_value_t regexp_value = ecma_copy_value_if_not_object (separator);
+          ecma_string_t *substr_str_p = ecma_get_magic_string (LIT_MAGIC_STRING__EMPTY);
+          ecma_value_t match_result;
+          match_result = ecma_regexp_exec_helper (regexp_value, ecma_make_string_value (substr_str_p), true);
+          should_return = !ecma_is_value_null (match_result);
+
+          ecma_deref_ecma_string (substr_str_p);
+          ecma_free_value (match_result);
+#else
+          return ecma_raise_type_error (ECMA_ERR_MSG ("REGEXP separator is disabled in split method."));
+#endif
         }
         else
         {
-          ECMA_TRY_CATCH (separator_to_string_val,
-                          ecma_op_to_string (arg1),
-                          ret_value);
+          ecma_string_t *separator_str_p = ecma_get_string_from_value (separator);
 
-          separator = ecma_copy_value (separator_to_string_val);
+          if (ecma_string_get_length (separator_str_p) == 0)
+          {
+            should_return = true;
+          }
+        }
 
-          ECMA_FINALIZE (separator_to_string_val);
+        if (!should_return)
+        {
+          /* 11.c */
+          ecma_string_t *zero_str_p = ecma_new_ecma_string_from_number (ECMA_NUMBER_ZERO);
+
+          ecma_value_t put_comp = ecma_builtin_helper_def_prop (new_array_p,
+                                                                zero_str_p,
+                                                                this_to_string_val,
+                                                                true,
+                                                                true,
+                                                                true,
+                                                                false);
+
+          JERRY_ASSERT (ecma_is_value_true (put_comp));
+          ecma_deref_ecma_string (zero_str_p);
         }
+      }
+      else /* if (string_length != 0) || !ecma_is_value_empty (ret_value) */
```

Comment:
This comment does not match. IMHO we should remove all of them, because it's hard to maintain them.